src/Features/LanguageServer/Protocol/ExternalAccess/Razor/SemanticTokensRangesHandler.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.ExternalAccess.Razor;
+
+[Method(SemanticRangesMethodName)]
+internal class SemanticTokensRangesHandler : ILspServiceRequestHandler<SemanticTokensRangesParams, SemanticTokens>
+{
+    private const int TokenSize = 5;
+    public const string SemanticRangesMethodName = "textDocument/semanticTokens/ranges";
+    private readonly IGlobalOptionService _globalOptions;
+    private readonly SemanticTokensRefreshQueue _semanticTokenRefreshQueue;
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public SemanticTokensRangesHandler(
+        IGlobalOptionService globalOptions,
+        SemanticTokensRefreshQueue semanticTokensRefreshQueue)
+    {
+        _globalOptions = globalOptions;
+        _semanticTokenRefreshQueue = semanticTokensRefreshQueue;
+    }
+
+    public async Task<SemanticTokens> HandleRequestAsync(
+            SemanticTokensRangesParams request,
+            RequestContext context,
+            CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(request.TextDocument, "TextDocument is null.");
+        if (request.Ranges.Length == 0)
+        {
+            return new SemanticTokens() { Data = Array.Empty<int>() };
+        }
+
+        var responseData = new List<int[]>();
```

Comment:
Should use a pooled array builder:

```suggestion
        using var _ = ArrayBuilder<int[]>.GetInstance(request.Ranges.Length, out var responseData);
```

Then change `StitchSemanticTokenResponsesTogether` to take an `int[][]` and call pass `responseData.ToArray()` into it